feat(nextjs-cdk-construct): add runtime options to regeneration

mathisobadia opened the pull request:
fixes issue #2048  
The regeneration function was the only function where the runtime could not be modified through setttings prop
This can cause the regeneration lambda function to time out an fail
The pull request adds the option to configure the regeneration runtime (memory, timeout, role, and node version)

as a note I have not modified the nomenclature for this function but it is the only lambda handler that is named as Function instead of Lambda for the others (nextApiLambda, nextDefaultLambda, nextImageLambda) but the regeneration lambda is named RegenerationFunction. I can modify that also if needed.

dmsolutionz commented on packages/serverless-components/nextjs-cdk-construct/src/index.ts:
```diff
@@ -116,11 +130,16 @@ export class NextJSLambdaEdge extends cdk.Construct {
+          role: this.edgeLambdaRole,
+          runtime:
+            toLambdaOption("regenerationLambda", props.runtime) ??
+            lambda.Runtime.NODEJS_12_X,
```
This should remain NODEJS_14_X?

mathisobadia replied:
I'm not 100% sure why it was hardcoded as NODEJS_14_X but I think it's unintuitive that the regeneration function is the only one with a fixed nodejs version. This might result in some weird bugs if it is the regeneration function uses a different runtime than the default function because they use the same js page bundles. This line I added is the same as for the other functions qnd it makes sure that all functions use the same nodejs version except if the user really wants to set something different throught props

mathisobadia replied:
Don't we want the regeneration lamda have the same role as the other functions? I know in my case I need this because the role grants access to dynamo to get the date necessary for generating a page

afenton90 replied:
Sorry @mathisobadia to be clear I think what @dmsolutionz is saying is that the "default should remain as NODEJS_14_X" rather than it shouldn't be editable. If you change to `NODEJS_12_X` as the default that's a breaking change.
Having it as editable like the others is exactly the desired outcome here.

mathisobadia replied:
Oh yeah that makes perfect sense! I've updated my pull request accordingly.

afenton90 reviewed (commented):
Looks good aside from the change mentioned by @dmsolutionz

mathisobadia commented:
> Looks good aside from the change mentioned by @dmsolutionz

Thank you for the feedback! I can change my pull request but I'm wondering why this function should be the only one with fixed node 14 runtime while all the other ones can be set with props? it doesn't make sense to keep the props as they are with the possibility to set the runtime version of the regeneration function since we now have 

```js
  /**
   * Lambda runtimes(s)
   */
  runtime?: LambdaOption<Runtime>;
```

and 

```js
export type LambdaOption<T> =
  | T
  | {
      defaultLambda?: T;
      apiLambda?: T;
      imageLambda?: T;
      regenerationLambda?: T;
    };
```

this means that the user can use a prop like 

```js
runtime: {
   regenerationLamda: Runtime.NODEJS_12_X
}
```

this would be valid typescript prop but it would not change the runtime of the function. This would be highly deceptive.
In my opinion if we keep the line as NODEJS_14_X then we also need to update the props so that it reflects the fact that the runtime is the only option that cannot be changed for the regeneration function. 

As an alternative, so that we don't change the runtime of current users unexpectedly we could keep the option to set the runtime but keep as a default the NODEJS_14_X instead of the default NODEJS_12_X that we have for all the other lambdas

mathisobadia commented:
I just noticed that the tests fail because of the snapshots I've updated the pull request to fix the put back the previous default memory (undefined) and remove the unnecessary role that I added by mistake
